### ISSUE TITLE
Close the precision gap: matrix-free sub-quadratic attention at 1.12×, can it reach 1.0×?

### ISSUE THREAD

**Issue by edisonbd** [edisonbd/cr-nn#1]
**Background.** This repo replaces softmax attention with matrix-free operators
(no $N\times N$ matrix). The best so far is an **LDR attention** — a Toeplitz
cross-correlation $Q\star K$ (two FFTs) plus a Linformer-style low-rank residual
($r=64$) — reaching **1.12×** of a Qwen2-style baseline (GQA+RoPE+flash) and
beating a same-scale GPT-2.

**What was tried (measured, 80M-token corpus, d=512, 4 layers, blockwise):**

| attention | eval ppl | rel. Qwen2 |
|---|---:|---:|
| complex Szegő (twist) | 1833 | 1.72× |
| real spectral (drop complex) | 1460 | 1.37× |
| $Q\star K$ cross-correlation | 1381 | 1.29× |
| **LDR ($Q\star K$ + low-rank $r{=}64$)** | **1201** | **1.12×** |
| low-rank alone (fixed pooling) | 1468 | 1.37× |
| $r{=}256$ low-rank | 1476 | overfits |
| stacked rounds $n_{flow}{=}2$ | 1819 | overfits |
| triple convolution $Q\star K\star V$ | explodes | pathological |
| gating $\sigma(W_g x)\odot v$ | no change | — |
| segmented per-segment kernels | 1681 | loses global interaction |

**The blocker.** The remaining gap is Toeplitz vs full matrix. Position-specific
capacity must be *small* ($r=64$ helps, $r=256$ overfits), so naively increasing
it fails.

**Reproduce.**
```bash
PYTHONPATH=. python experiments/03_toy_seq/ldr_attn.py --d 512 --layers 4 --p 17 --r 64
```

**Help wanted.** Any *matrix-free* sub-quadratic interaction that gets below
1.12×. Candidates not yet explored: learned (not strided) low-rank projections,
multi-scale cross-correlation, input-adaptive kernel selection (Mamba-style
selectivity).

---
